src/Dapr.Client/DaprClientGrpc.cs

```diff
@@ -210,124 +197,177 @@ private static T ConvertFromInvokeBindingResponse<T>(Autogenerated.InvokeBinding
+
+            if (!this.httpEndpoint.IsBaseOf(request.RequestUri))
             {
-                serializedData = TypeConverters.ToAny(data, this.jsonSerializerOptions);
+                throw new InvalidOperationException("The provided request URI is not a Dapr service invocation URI.");
             }
 
-            Autogenerated.InvokeServiceRequest request;
-            CallOptions callOptions;
-            (request, callOptions) = this.MakeInvokeRequestAsync(appId, methodName, serializedData, httpOptions, cancellationToken);
-            await client.InvokeServiceAsync(request, callOptions);
+            // Note: we intentionally DO NOT validate the status code here.
+            // This method allows you to 'invoke' without exceptions on non-2xx.
+            return this.httpClient.SendAsync(request, cancellationToken);
         }
 
-        public override async Task<TResponse> InvokeMethodAsync<TResponse>(
-           string appId,
-           string methodName,
-           HttpInvocationOptions httpOptions = default,
-           CancellationToken cancellationToken = default)
+        public async override Task InvokeMethodAsync(HttpRequestMessage request, CancellationToken cancellationToken = default)
+        {
+            ArgumentVerifier.ThrowIfNull(request, nameof(request));
+
+            var response = await InvokeMethodWithResponseAsync(request, cancellationToken);
+            response.EnsureSuccessStatusCode();
+        }
+
+        public async override Task<TResponse> InvokeMethodAsync<TResponse>(HttpRequestMessage request, CancellationToken cancellationToken = default)
+        {
+            ArgumentVerifier.ThrowIfNull(request, nameof(request));
+
+            var response = await InvokeMethodWithResponseAsync(request, cancellationToken);
+            response.EnsureSuccessStatusCode();
+            return await response.Content.ReadFromJsonAsync<TResponse>(this.jsonSerializerOptions, cancellationToken);
+        }
+
+        public override async Task InvokeMethodGrpcAsync(string appId, string methodName, CancellationToken cancellationToken = default)
         {
             ArgumentVerifier.ThrowIfNullOrEmpty(appId, nameof(appId));
             ArgumentVerifier.ThrowIfNullOrEmpty(methodName, nameof(methodName));
 
-            Autogenerated.InvokeServiceRequest request;
-            CallOptions callOptions;
-            (request, callOptions) = this.MakeInvokeRequestAsync(appId, methodName, null, httpOptions, cancellationToken);
-            var response = await client.InvokeServiceAsync(request, callOptions);
-            return response.Data.Value.IsEmpty ? default : TypeConverters.FromAny<TResponse>(response.Data, this.jsonSerializerOptions);
+            var envelope = new Autogenerated.InvokeServiceRequest()
+            {
+                Id = appId,
+                Message = new Autogenerated.InvokeRequest()
+                {
+                    Method = methodName,
```

Comment:
nit: extra whitespace

Reply:
I don't see any, I autoformatted the document and it's still in the same place.